storage/innobase/srv/srv0start.cc

```diff
@@ -1988,6 +1985,20 @@ void innodb_preshutdown()
 
   if (srv_read_only_mode)
     return;
+
+#ifndef EMBEDDED_LIBRARY
+  if (srv_force_recovery < SRV_FORCE_NO_TRX_UNDO && srv_was_started &&
+      !trx_rollback_is_active && trx_sys.any_active_transaction_recovered())
+  {
+    /* Trx rollback thread may not be started yet, since error happened before
+       innobase_tc_log_recovery_done(), thus it is started here to finish the
+       normal shutdown process.
+     */
+    trx_rollback_is_active= true;
+    srv_thread_pool->submit_task(&rollback_all_recovered_task);
```

Comment:
Like you pointed out, we do need this for any value of `innodb_fast_shutdown`, because the recovered transactions must be removed from `trx_sys` by the rollback task, even if the rollback itself is aborted. But a message is not needed here. Thank you for removing it.